### DISCUSSION TITLE
Polaris units

### DISCUSSION THREAD

**Discussion by lgriffee** [Shopify/polaris-react#5566] [Q&A]
- Do we want to create an underlying set of Polaris units? 
  > From @alex-page: 
  > `polaris units` all the units from the system
  > `semantic tokens` spacing, type, line width, radius using the system units

- If so what should the values be and where should they be used?

**Comment by alex-page** [comment]
This could be a terrible idea but this is my thinking.

**Polaris units**
all the values provided from the system. As a user I can use these for creating new semantic relationships and stay on mainline.

**Fontsize tokens**
all of the values that are available for typography. There isn’t anything smart here just a list of values.

**Components**
Use the token values and change them at different break points.

Polaris units become a central source of truth of all the units that we scale from the system. We semantically apply them to different values for border-radius, font-size, spacing etc. These semantic values are then used to build responsive components that could use multiple tokens.

![Screen Shot 2022-04-20 at 10 19 42 AM](https://user-images.githubusercontent.com/19199063/164287760-fd54a2ea-a148-46bf-b463-a517d30cbaec.png)

**Reply by alex-page** [reply]
The units could be based off a scale. Or we could create the units from a scale for typography and push them into the polaris units. I don't think this needs to be a right to left diagram.

**Reply by heyjoethomas** [reply]
My first thought is this seems fairly over-engineered for what doesn't seem like a massive problem or value add. I'm struggling to envision how this makes our lives easier than a more simple approach where we deal with each thing individually (type scale, spacing scale, size scale, radius scale, etc). That seems far more legible & maintainable.

**Reply by alex-page** [reply]
If we have many scales they are still in some way related to each other through the user interface. How would you strengthen the relationships between multiple scales?

If I am doing a `top` value do I take from `spacing` or do I take from something else? Do we create a new scale? Is there a list of units that could connect someone to the mainline when the individual layers don't?

**Reply by heyjoethomas** [reply]
Sure they are related and so should be constructed thoughtfully and with each other in mind. I guess I just don't think we should create another layer that puts them together. The only multiple scale we have right now is typography which I know Sara is looking to see if we even need or if we could do with a single scale. 

If you need to add space you would take from the space scale. If you want to define the size of something, use the sizing scale (if we have one). I would imagine either could be applied via prop or via a design token depending on circumstance. But either would be on mainline. Seems very intuitive to me 🤷

**Reply by alex-page** [reply]
If it is all connected we could think of this layer like units `px` `em` and `rem`. It could radically simplify how we do units across layers of tokens. We could have a set group of units for polaris for example `1pu` and that could be the link.

> The only multiple scale we have right now is typography

Tokens shouldn't come in different scales. This approach caused a lot of confusion for developers and stopped them from using the values all together.

**Reply by sarahill** [reply]
> Tokens shouldn't come in different scales. This approach caused a lot of confusion for developers and stopped them from using the values all together.

If we choose to adjust font-sizes on mobile I don't know that we can avoid different scales completely. The needs may just be different so I don't really want to completely rule out two scales. 

If, for example,  `--heading-lg` on desktop got a value of `20px` and on mobile the value changed to `18px` (values just for example) automatically based on a breakpoint, that feels expected. Do you think that causes confusion?

**Reply by alex-page** [reply]
> If we choose to adjust font-sizes on mobile I don't know that we can avoid different scales completely. The needs may just be different so I don't really want to completely rule out two scales.

That is all in the component. That has nothing to do with tokens. We shouldn't be having opinions on how tokens change on the token layer. The components can decide that.

`<Heading size="large">` can use the `--font-size-14` on desktop and `--font-size-18` on mobile.

**Reply by aaronccasanova** [reply]
@lgriffee and I put together a rough demo of the proposed units, type units, etc. scales to better understand how this could work in practice: https://codepen.io/aaronccasanova/pen/zYpbvqV

We had a similar reaction to @heyjoethomas and didn't feel the complexity added much value. After creating the example we are still not entirely sold on the approach, but at least have a better understanding of how we could use interrelated unit scales.

**Reply by heyjoethomas** [reply]
1. If we zoom out, the thing we don't want (in @sarahill and I's opinion) is all text sizes available all the time. Sizes that were specifically chosen to only appear on mobile, next to other mobile specific sizes, shouldn't be able to be used on desktop. That is the worry a single scale brings. That is how our docs read now and it is super confusing. Going to that page now makes it seems like I can use any/all of those values whenever I want...which isn't true. That is the concern imo.
2. @aaronccasanova That was super helpful to see. Thank you. My question would be this. Is this a high impact need worth solving right now? Are we really wanting/envisioning shipping the ability to change the entire density of Admin? That seems to me what this idea is suggesting and I would genuinely wonder if that is a real/big/impactful problem or more of a "nice-to-have". 

Also, this is a fantastic discussion 🤜 🤛

**Reply by alex-page** [reply]
> Sizes that were specifically chosen to only appear on mobile, next to other mobile specific sizes, shouldn't be able to be used on desktop. That is the worry a single scale brings.

This is how font sizes worked previously in scss. It makes the typography tokens more complex as they are no longer single name and value relationships. `--font-size-heading` being two different values means that users need to know breakpoints to use them and that they have have to dig into documentation to see the different sizes.

This approach lead to lots of teams not using the system. As a developer I might not want the relationship defined by the system and now I would be forced to hard code something. I think it's much better to just have a scale and let the components set the rules for when specific sizes should and should not be used.

> Going to that page now makes it seems like I can use any/all of those values whenever I want...which isn't true. That is the concern imo.

This is exactly what we need. Components should have the opinions not the tokens. We need to enable folks to use any values in font scale at all breakpoints. Lots of trust.

**Reply by sarahill** [reply]
It sounds like you are against composite tokens, is that right? If we didn't have that layer, how would we approach sensible defaults? or is that layer just at the component level and not exposed elsewhere?

**Reply by alex-page** [reply]
I am against tokens being multiple values based on some decision. Sensible defaults for combinations of fontsizes I would put at the component layer. `--font-size-heading` is just `<Heading>`.

**Reply by sarahill** [reply]
We chatted about that a bit yesterday in context of the text components. I'm on board with `<Heading />` being the sensible default along with things like` <Heading level="h3" style="2xl" />` for the other defaults and then adding more flexibility by letting users have complete control with setting size/style. I'll leave these details for Emma's discussion though in terms of how a component(s) might work.

My brain still can't quite understand how we avoid the two values for tools like Figma but something we can dig into.

**Reply by alex-page** [reply]
I would love to learn about the restrictions in Figma can we huddle on this and post an update here?

**Reply by sarahill** [reply]
Sure. As I think about it, we could make Figma components for the sensible defaults with breakpoint toggles for how the values change within the component instead of making separate figma text styles. cc: @heyjoethomas

**Reply by heyjoethomas** [reply]
@alex-page  @sarahill @emmaorhun  We should def huddle/jam on this when Sara is feeling better. Feel like we have multiple ideas floating around lol. 

In Figma text styles can only have 1 definition. That definition contains the following attributes:

- font-family
- weight
- size
- line-height
- text decoration

Once you change any of those attributes you need a new text style. Also, there are no variants like with components where we can toggle between large-screen and small-screen. They have to be separately defined and names styles.

**Reply by aaronccasanova** [reply]
@heyjoethomas Please add me to this huddle 🙏
